### PR TITLE
 Sql Implementation for Add custom tag

### PR DESCRIPTION
**Pull request by pengchen0692**
## Description
Implement SQL part for AddCustomTag. Now we can see effect in database.

## Related issues
Addresses [AB#79175](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/79175)

## Testing
* Verified calling add customtag API do insert row in database
* Verified schemaV1 is automatically upgraded to V2
* Verified on new machine, database is created as V2.
*

### CODE REVIEW

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/1.sql [hunk 1]
```diff
@@ -473,10 +473,8 @@ AS
         AND SopInstanceUid = @sopInstanceUid
 
     IF @@ROWCOUNT <> 0
-    BEGIN
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/1.sql:476]: Revert these changes?

File: src/Microsoft.Health.Dicom.Core/Extensions/CustomTagEntryExtensions.cs [hunk 2]
```diff
@@ -47,7 +47,7 @@ public static CustomTagEntry Normalize(this CustomTagEntry customTagEntry)
 
             vr = vr.ToUpperInvariant();
 
-            return new CustomTagEntry(path, vr, customTagEntry.Level, customTagEntry.Status);
+            return new CustomTagEntry() { Path = path, VR = vr, Level = customTagEntry.Level, Status = customTagEntry.Status };
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Extensions/CustomTagEntryExtensions.cs:50]: nit: you don't need the `()` for the ctor

File: src/Microsoft.Health.Dicom.Core/Extensions/DicomTagExtensions.cs [hunk 3]
```diff
@@ -28,7 +28,7 @@ public static string GetPath(this DicomTag dicomTag)
         /// <summary>
         /// Get default VR for dicom tag.
         /// </summary>
-        /// <remarks>If the dicom tag is private or unknown tag, Null is returned.</remarks>
+        /// <remarks>If the dicom tag is private or unknown tag, null is returned.</remarks>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Extensions/DicomTagExtensions.cs:31]: nit: `<see langword="null" />`

File: src/Microsoft.Health.Dicom.Core/Modules/ServiceModule.cs [hunk 4]
```diff
@@ -110,7 +111,7 @@ public void Load(IServiceCollection services)
                 .AsImplementedInterfaces();
 
             services.Add<DicomTagParser>()
-                         .Scoped()
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Modules/ServiceModule.cs:113]: nit: can you reduce the indent here to only 1 level?

```csharp
services.Add<DicomTagParser>()
    .Singleton()
    .AsSelf()
    .AsImplementedInterfaces();
```

File: src/Microsoft.Health.Dicom.Core/Modules/ServiceModule.cs [hunk 5]
```diff
@@ -110,7 +111,7 @@ public void Load(IServiceCollection services)
...
+                         .Singleton()
                          .AsSelf()
                          .AsImplementedInterfaces();
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Modules/ServiceModule.cs:116]: I am unfamiliar with these methods -- I'm looking at them now, but based on their names and usage I am assuming they are probably equivalent to the following (as long as the class only implements 1 interface and we don't access `DicomTagParser` through the concrete type).

```csharp
services.AddSingleton<IDicomTagParser, DicomTagParser>();
```

File: src/Microsoft.Health.Dicom.Core.UnitTests/Extensions/CustomTagEntryExtensionsTests.cs [hunk 6]
```diff
@@ -78,7 +78,12 @@ public void GivenStandardTagAsKeyword_WhenFormalizing_ThenVRShouldBeFilled()
         public static IEnumerable<object[]> GetValidCustomTagEntries()
         {
             yield return new object[] { DicomTag.DeviceSerialNumber.BuildCustomTagEntry() }; // standard custom tag with VR
-            yield return new object[] { new CustomTagEntry("12051003", DicomVRCode.OB, CustomTagLevel.Instance, CustomTagStatus.Added) }; // private tag with VR
+            yield return new object[] { CreateCustomTagEntry("12051003", DicomVRCode.OB, CustomTagLevel.Instance, CustomTagStatus.Added) }; // private tag with VR
+        }
+
+        private static CustomTagEntry CreateCustomTagEntry(string path, string vr, CustomTagLevel level = CustomTagLevel.Instance, CustomTagStatus status = CustomTagStatus.Added)
+        {
+            return new CustomTagEntry() { Path = path, VR = vr, Level = level, Status = status };
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core.UnitTests/Extensions/CustomTagEntryExtensionsTests.cs:86]: nit: remove the `()` for the ctor

File: src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/CustomTagEntryValidatorTests.cs [hunk 7]
```diff
@@ -123,7 +123,7 @@ public void GivenDuplicatedTag_WhenValidating_ThenShouldThrowException()
 
         private static CustomTagEntry CreateCustomTagEntry(string path, string vr, CustomTagLevel level = CustomTagLevel.Instance, CustomTagStatus status = CustomTagStatus.Added)
         {
-            return new CustomTagEntry(path, vr, level, status);
+            return new CustomTagEntry() { Path = path, VR = vr, Level = level, Status = status };
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/CustomTagEntryValidatorTests.cs:126]: nit: remove the `()`

File: src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/DicomTagExtensions.cs [hunk 8]
```diff
@@ -13,7 +13,7 @@ public static class DicomTagExtensions
     {
         public static CustomTagEntry BuildCustomTagEntry(this DicomTag tag, CustomTagLevel level = CustomTagLevel.Series, CustomTagStatus status = CustomTagStatus.Added)
         {
-            return new CustomTagEntry(path: tag.GetPath(), vr: tag.GetDefaultVR()?.Code, level: level, status: status);
+            return new CustomTagEntry() { Path = tag.GetPath(), VR = tag.GetDefaultVR()?.Code, Level = level, Status = status };
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core.UnitTests/Features/CustomTag/DicomTagExtensions.cs:16]: nit: remove the `()`

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 9]
```diff
@@ -923,3 +923,93 @@ BEGIN
...
+    MINVALUE 1
+    NO CYCLE
+    CACHE 10000
+GO
+
+/*************************************************************
+    USER DEFINED TABLES
+*************************************************************/
+/*************************************************************
+    The user defined table for AddCustomTagsInput
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:975]: nit: type not a table

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 10]
```diff
@@ -923,3 +923,93 @@ BEGIN
...
+*************************************************************/
+CREATE TYPE dbo.AddCustomTagsInputTableType_1 AS TABLE
+(
+    TagPath                    VARCHAR(64),  -- Custom Tag Path. Each custom tag take 8 bytes, support upto 8 levels, no delimeter between each level.
+    TagVR                      VARCHAR(2),  -- Custom Tag VR.
+    TagLevel                   TINYINT,  -- Custom Tag level. 0 -- Instance Level, 1 -- Series Level, 2 -- Study Level
+    TagStatus                  TINYINT
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:982]: Can you assume the status in the procedure, Since you are just adding.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 11]
```diff
@@ -923,3 +923,93 @@ BEGIN
...
+
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+
+    DECLARE @duplicateCount BIGINT
+
+    BEGIN TRANSACTION
+        
+        -- Check if tag with same path already exist
+        SELECT TagKey FROM dbo.CustomTag INNER JOIN @customTags input ON input.TagPath = dbo.CustomTag.TagPath
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1003]: nit: format it to similar queries above.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 12]
```diff
@@ -923,3 +923,93 @@ BEGIN
...
+	    
+        SET @duplicateCount = @@ROWCOUNT
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1005]: Why not just use @@ROWCOUNT, do we need the variable?

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 13]
```diff
@@ -923,3 +923,93 @@ BEGIN
...
+    TagLevel                TINYINT              NOT NULL,
+    TagStatus               TINYINT              NOT NULL,
+)
+
+CREATE UNIQUE CLUSTERED INDEX IXC_CustomTag ON dbo.CustomTag
+(
+    TagKey
+)
+
+CREATE NONCLUSTERED INDEX IX_CustomTag_TagPath ON dbo.CustomTag
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:953]: can we do a unique on this this index?

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 14]
```diff
@@ -923,3 +923,93 @@ BEGIN
...
+AS
+
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+
+    DECLARE @duplicateCount BIGINT
+
+    BEGIN TRANSACTION
+        
+        -- Check if tag with same path already exist
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1002]: You may want to think about concurrent transactions and SQL locking. 

WITH(HOLDLOCK) on the select and unique index on the tagPath will stop any duplicate tagpaths getting added from concurrent AddCustomTags

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/SchemaVersion.cs [hunk 15]
```diff
@@ -11,5 +11,6 @@ namespace Microsoft.Health.Dicom.SqlServer.Features.Schema
     public enum SchemaVersion
     {
         V1 = 1,
+        V2 = 2,
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/SchemaVersion.cs:14]: I know we discussed this in the standup. Are we keeping this in a Feature branch untill we have initial schema of everything? Since this is OSS now, we need to make sure users can upgrade.

File: src/Microsoft.Health.Dicom.Core.UnitTests/Extensions/CustomTagEntryExtensionsTests.cs (hunk 6 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core.UnitTests/Extensions/CustomTagEntryExtensionsTests.cs:86]: never know this. :)

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql (hunk 13 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:953]: I guess Sneha will do this, but it's also good to change here. Updated.

File: src/Microsoft.Health.Dicom.Core.UnitTests/Extensions/CustomTagEntryExtensionsTests.cs (hunk 6 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core.UnitTests/Extensions/CustomTagEntryExtensionsTests.cs:86]: A bit of a shortcut lol. It is only valid for this object initializer syntax

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql (hunk 14 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1002]: learning SQL locking and make this propose, please check if it's the right way.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/SchemaVersion.cs (hunk 15 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/SchemaVersion.cs:14]: The plan is to keep this on feature branch until the min scope is completed. Then merge to main branch.
min scope include: Get/List/Add/Delete custom tag(s), support custom tag on QIDO, STOW. Without the reindex feature.

We can talk if you have concerns.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql (hunk 12 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1005]: There is syntax error when with "@duplicateCount".

Another think:  I use state to pass duplicated row count back, is this the right usage? I see we normally return 1, how about I do same thing here?

File: src/Microsoft.Health.Dicom.Core/Modules/ServiceModule.cs (hunk 4 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Modules/ServiceModule.cs:113]: If we use the same pattern (of `AsSelf()` and `AsImplementedInterfaces()`) everywhere, then it's probably worth a conversation with the team about how we declare these services. So I would leave this with your previous usage of:

```csharp
services.Add<DicomTagParser>()
    .Singleton()
    .AsSelf()
    .AsImplementedInterfaces();
```

If we want to change it, then we can change all of them

File: src/Microsoft.Health.Dicom.SqlServer/Features/CustomTag/SqlCustomTagStore.cs [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            if (_schemaInformation.Current < SchemaVersionConstants.SupportCustomTagSchemaVersion)
+            {
+                throw new BadRequestException(DicomSqlServerResource.SchemaVersionNeedsToBeUpgraded);
+            }
+
+            using (SqlConnectionWrapper sqlConnectionWrapper = await _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                IEnumerable<AddCustomTagsInputTableTypeV1Row> rows = customTagEntries.Select(entry => ToAddCustomTagsInputTableTypeV1Row(entry));
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/CustomTag/SqlCustomTagStore.cs:54]: nit: you can use the static method directly as an argument (converted to a delegate implicitly):

```csharp
customTagEntries.Select(ToAddCustomTagsInputTableTypeV1Row);
```

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 17]
```diff
@@ -923,3 +923,100 @@ BEGIN
...
+        INNER JOIN @customTags input 
+        ON input.TagPath = dbo.CustomTag.TagPath 
+	    
+        IF @@ROWCOUNT <> 0
+            THROW 50409, 'custom tag(s) already exist', 1 
+
+        -- add to custom tag table with status 1(Added)
+        INSERT INTO dbo.CustomTag 
+            (TagKey, TagPath, TagVR, TagLevel, TagStatus)
+        SELECT NEXT VALUE FOR TagKeySequence, TagPath, TagVR,TagLevel, 1 FROM @customTags
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1019]: nit: missing space before `TagLevel`

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql [hunk 18]
```diff
@@ -923,3 +923,100 @@ BEGIN
...
+    BEGIN TRANSACTION
+        
+        -- Check if tag with same path already exist
+        SELECT TagKey 
+        FROM dbo.CustomTag WITH(HOLDLOCK) 
+        INNER JOIN @customTags input 
+        ON input.TagPath = dbo.CustomTag.TagPath 
+	    
+        IF @@ROWCOUNT <> 0
+            THROW 50409, 'custom tag(s) already exist', 1
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1014]: We use embedded resource files for our assemblies today in .NET which can use the current culture to select the most appropriate error message. We may wish to do something similar for our SQL error messages as we add stored procedures. We can add messages for different locales with `sp_addmessage`. See [here](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/throw-transact-sql?view=sql-server-ver15#c-using-formatmessage-with-throw)

I am unsure though if this is required for this PR. It probably makes sense to evaluate this in another PR.

File: src/Microsoft.Health.Dicom.SqlServer/Microsoft.Health.Dicom.SqlServer.csproj [hunk 19]
```diff
@@ -21,18 +25,40 @@
   <ItemGroup>
     <ProjectReference Include="..\Microsoft.Health.Dicom.Core\Microsoft.Health.Dicom.Core.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Features\CustomTag\" />
+    <Folder Include="Features\Schema\Model\" />
+  </ItemGroup>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Microsoft.Health.Dicom.SqlServer.csproj:31]: What are these `<Folder/>` elements for?

File: src/Microsoft.Health.Dicom.Core/DicomCoreResource.resx [hunk 20]
```diff
@@ -299,4 +295,7 @@ The first part date {1} should be lesser than or equal to the second part date {
     <value>The CR code '{0}' is not supported.</value>
     <comment>{0} VR Code.</comment>
   </data>
+  <data name="CustomTagsAlreadyExists" xml:space="preserve">
+    <value>One or multiple custom tag(s) already exist(s).</value>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/DicomCoreResource.resx:299]: nit: `One or more custom tags already exist.`

**Review by wsugarman — APPROVED** [review]

**Review by smithago — APPROVED** [review]

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql (hunk 18 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.sql:1014]: may have separate PR to update all such string.

File: src/Microsoft.Health.Dicom.SqlServer/Microsoft.Health.Dicom.SqlServer.csproj (hunk 19 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Microsoft.Health.Dicom.SqlServer.csproj:31]: It's added by Visual Studio, let me see if can get rid of.
